cmd/subctl/join.go

```diff
@@ -55,34 +58,31 @@ var joinCmd = &cobra.Command{
 		exit.OnError(status.Error(err, "Error loading the broker information from the given file"))
 		status.Success("%s indicates broker is at %s", args[0], brokerInfo.BrokerURL)
 
-		determineClusterID(status)
-
-		clientConfig, err := restConfigProducer.ForCluster()
-		exit.OnError(status.Error(err, "Error creating the REST config"))
+		exit.OnError(joinRestConfigProducer.RunOnSelectedContext(
+			func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
+				determineClusterID(clusterInfo.Name, status)
 
-		clientProducer, err := client.NewProducerFromRestConfig(clientConfig.Config)
-		exit.OnError(status.Error(err, "Error creating the client producer"))
+				networkDetails := getNetworkDetails(clusterInfo.ClientProducer, status)
+				determinePodCIDR(networkDetails, status)
+				determineServiceCIDR(networkDetails, status)
 
-		networkDetails := getNetworkDetails(clientProducer, status)
-		determinePodCIDR(networkDetails, status)
-		determineServiceCIDR(networkDetails, status)
+				if brokerInfo.IsConnectivityEnabled() && labelGateway {
+					possiblyLabelGateway(clusterInfo.ClientProducer.ForKubernetes(), status)
+				}
 
-		if brokerInfo.IsConnectivityEnabled() && labelGateway {
-			possiblyLabelGateway(clientProducer.ForKubernetes(), status)
-		}
-
-		if joinFlags.CustomDomains == nil && brokerInfo.CustomDomains != nil {
-			joinFlags.CustomDomains = *brokerInfo.CustomDomains
-		}
+				if joinFlags.CustomDomains == nil && brokerInfo.CustomDomains != nil {
+					joinFlags.CustomDomains = *brokerInfo.CustomDomains
+				}
 
-		err = join.ClusterToBroker(brokerInfo, &joinFlags, clientProducer, status)
-		exit.OnError(err)
+				return join.ClusterToBroker( // nolint:wrapcheck // No need to wrap errors here.
+					brokerInfo, &joinFlags, clusterInfo.ClientProducer, status)
```

Comment:
I find it relatively hard to read this anonymous function with such a high level of indentation.
Do you think we can extract it to a named function, for easier code maintainability?

Reply:
Theoretically you could get an ellipsis or args as optional last argument to the `RunOnSelectedContext` function and pass it on to the nested function, but it might only serve this one use case, while complicating the function signature for everyone :thinking: